### PR TITLE
Throw error when identity in both url and token

### PR DESCRIPTION
**Pull request by tzheleznyak**
Throw error when identity when identity given in URL and also "token-app-proprery" configured

### CODE REVIEW

File: app/domain/errors.rb [hunk 1]
```diff
@@ -474,6 +474,16 @@ module AuthnJwt
         code: "CONJ00097E"
       )
 
+      ServiceIdMissing = ::Util::TrackableErrorClass.new(
+        msg: "Service ID is required when authenticating with authn-jwt",
+        code: "CONJ00097E"
+      )
+
+      IdentityMisconfigured = ::Util::TrackableErrorClass.new(
```
Comment (tzheleznyak) [inline app/domain/errors.rb:482]: Hey @shulifink can you review this error message
CC : @hilagross

File: app/domain/authentication/authn_jwt/identity_providers/create_identity_provider.rb [hunk 2]
```diff
@@ -22,16 +22,22 @@ def call
...
         def create_identity_provider
-          identity_provider = @identity_from_decoded_token_class.new(@authentication_parameters)
-          if identity_provider.identity_available?
+          identity_from_decoded_token_provider = @identity_from_decoded_token_class.new(@authentication_parameters)
+          identity_from_url_provider = @identity_from_url_provider_class.new(@authentication_parameters)
+
+          if identity_from_decoded_token_provider.identity_available and identity_from_url_provider.identity_available
+            raise Errors::Authentication::AuthnJwt::IdentityMisconfigured
+          end
+
```
Comment (Avi-Eyal) [inline app/domain/authentication/authn_jwt/identity_providers/create_identity_provider.rb:31]: remove blank line

File: app/domain/authentication/authn_jwt/identity_providers/create_identity_provider.rb [hunk 3]
```diff
@@ -22,16 +22,22 @@ def call
...
+          if identity_from_decoded_token_provider.identity_available
             @logger.debug(LogMessages::Authentication::AuthnJwt::DECODED_TOKEN_IDENTITY_PROVIDER_SELECTED.new)
-            return identity_provider
+            return identity_from_decoded_token_provider
           end
 
-          identity_provider = @identity_from_url_provider_class.new(@authentication_parameters)
-          if identity_provider.identity_available?
+          if identity_from_url_provider.identity_available
```
Comment (Avi-Eyal) [inline app/domain/authentication/authn_jwt/identity_providers/create_identity_provider.rb:38]: why isnt it "elsif"?

File: app/domain/authentication/authn_jwt/identity_providers/create_identity_provider.rb [hunk 4]
```diff
@@ -22,16 +22,22 @@ def call
...
+          identity_from_decoded_token_provider = @identity_from_decoded_token_class.new(@authentication_parameters)
+          identity_from_url_provider = @identity_from_url_provider_class.new(@authentication_parameters)
+
+          if identity_from_decoded_token_provider.identity_available and identity_from_url_provider.identity_available
+            raise Errors::Authentication::AuthnJwt::IdentityMisconfigured
+          end
+
+
+          if identity_from_decoded_token_provider.identity_available
             @logger.debug(LogMessages::Authentication::AuthnJwt::DECODED_TOKEN_IDENTITY_PROVIDER_SELECTED.new)
```
Comment (Avi-Eyal) [inline app/domain/authentication/authn_jwt/identity_providers/create_identity_provider.rb:34]:
```suggestion
            @logger.info(LogMessages::Authentication::AuthnJwt::DECODED_TOKEN_IDENTITY_PROVIDER_SELECTED.new)
```

like Nessi's changes

File: app/domain/authentication/authn_jwt/identity_providers/create_identity_provider.rb [hunk 5]
```diff
@@ -22,16 +22,22 @@ def call
...
-            return identity_provider
+            return identity_from_decoded_token_provider
           end
 
-          identity_provider = @identity_from_url_provider_class.new(@authentication_parameters)
-          if identity_provider.identity_available?
+          if identity_from_url_provider.identity_available
             @logger.debug(LogMessages::Authentication::AuthnJwt::URL_IDENTITY_PROVIDER_SELECTED.new)
```
Comment (Avi-Eyal) [inline app/domain/authentication/authn_jwt/identity_providers/create_identity_provider.rb:39]:
```suggestion
            @logger.info(LogMessages::Authentication::AuthnJwt::URL_IDENTITY_PROVIDER_SELECTED.new)
```
here too

File: app/domain/authentication/authn_jwt/identity_providers/identity_from_url_provider.rb [hunk 6]
```diff
@@ -8,12 +8,16 @@ def initialize(authentication_parameters)
...
 
           @authentication_parameters.username
         end
 
-        def identity_available?
+        def identity_available
+          @identity_available ||= !@authentication_parameters.username.blank?
+        end
+
+        def check_username_in_url
```
Comment (Avi-Eyal) [inline app/domain/authentication/authn_jwt/identity_providers/identity_from_url_provider.rb:20]: its not really checking, maybe call it `username_exists?` ?

**Review by Avi-Eyal — CHANGES_REQUESTED** [review]
Please change

File: app/domain/errors.rb [hunk 7]
```diff
@@ -474,6 +474,16 @@ module AuthnJwt
         code: "CONJ00097E"
       )
 
+      ServiceIdMissing = ::Util::TrackableErrorClass.new(
+        msg: "Service ID is required when authenticating with authn-jwt",
```
Comment (benoaviram) [inline app/domain/errors.rb:478]: Duplicated by mistake, remove it please

File: app/domain/authentication/authn_jwt/identity_providers/identity_from_decoded_token_provider.rb [hunk 8]
```diff
@@ -28,14 +28,14 @@ def jwt_identity
         end
 
         # Checks if variable that defined from which field in decoded token to get the id is configured
-        def identity_available?
-          identity_field_variable.present?
+        def identity_available
+          @identity_available ||= identity_field_variable.present?
```
Comment (Avi-Eyal) [inline app/domain/authentication/authn_jwt/identity_providers/identity_from_decoded_token_provider.rb:32]: fix memoization to new memoization (with unless)
In the current memoization if `identity_field_variable.present?` equals `false`, `identity_field_variable.present?` will be called again (it will not be memoized)

File: app/domain/authentication/authn_jwt/identity_providers/identity_from_url_provider.rb [hunk 9]
```diff
@@ -8,12 +8,16 @@ def initialize(authentication_parameters)
...
         def jwt_identity
-          raise Errors::Authentication::AuthnJwt::NoRelevantIdentityProvider unless identity_available?
+          raise Errors::Authentication::AuthnJwt::NoRelevantIdentityProvider unless identity_available
 
           @authentication_parameters.username
         end
 
-        def identity_available?
+        def identity_available
+          @identity_available ||= !@authentication_parameters.username.blank?
```
Comment (Avi-Eyal) [inline app/domain/authentication/authn_jwt/identity_providers/identity_from_url_provider.rb:17]: same problem with memoization

File: app/domain/authentication/authn_jwt/identity_providers/create_identity_provider.rb (hunk 2 above)
Comment (tzheleznyak) [reply app/domain/authentication/authn_jwt/identity_providers/create_identity_provider.rb:31]: After all if parts we need this line

File: app/domain/authentication/authn_jwt/identity_providers/create_identity_provider.rb (hunk 3 above)
Comment (tzheleznyak) [reply app/domain/authentication/authn_jwt/identity_providers/create_identity_provider.rb:38]: Done

File: app/domain/authentication/authn_jwt/identity_providers/create_identity_provider.rb (hunk 5 above)
Comment (tzheleznyak) [reply app/domain/authentication/authn_jwt/identity_providers/create_identity_provider.rb:39]: Nessi changing the logs

File: app/domain/authentication/authn_jwt/identity_providers/create_identity_provider.rb [hunk 10]
```diff
@@ -22,16 +22,22 @@ def call
         private
 
         def create_identity_provider
-          identity_provider = @identity_from_decoded_token_class.new(@authentication_parameters)
-          if identity_provider.identity_available?
+          identity_from_decoded_token_provider = @identity_from_decoded_token_class.new(@authentication_parameters)
+          identity_from_url_provider = @identity_from_url_provider_class.new(@authentication_parameters)
+
+          if identity_from_decoded_token_provider.identity_available and identity_from_url_provider.identity_available
```
Comment (nessiLahav) [inline app/domain/authentication/authn_jwt/identity_providers/create_identity_provider.rb:28]: what about they both unavailable ?

File: app/domain/authentication/authn_jwt/identity_providers/create_identity_provider.rb (hunk 4 above)
Comment (tzheleznyak) [reply app/domain/authentication/authn_jwt/identity_providers/create_identity_provider.rb:31]: This be done in nessi pr

File: app/domain/errors.rb [hunk 11]
```diff
@@ -474,6 +474,16 @@ module AuthnJwt
...
+        code: "CONJ00097E"
+      )
+
+      IdentityMisconfigured = ::Util::TrackableErrorClass.new(
+        msg: "Identity not configured correctly",
```
[the remark this answers is not in the document]
Comment (tzheleznyak) [reply app/domain/errors.rb:478]: Yes

File: app/domain/authentication/authn_jwt/identity_providers/create_identity_provider.rb (hunk 10 above)
Comment (tzheleznyak) [reply app/domain/authentication/authn_jwt/identity_providers/create_identity_provider.rb:28]: We throw error

File: app/domain/authentication/authn_jwt/identity_providers/identity_from_url_provider.rb [hunk 12]
```diff
@@ -8,12 +8,17 @@ def initialize(authentication_parameters)
...
         def jwt_identity
-          raise Errors::Authentication::AuthnJwt::NoRelevantIdentityProvider unless identity_available?
+          raise Errors::Authentication::AuthnJwt::NoRelevantIdentityProvider unless identity_available
 
           @authentication_parameters.username
         end
 
-        def identity_available?
+        def identity_available
+          @identity_available ||= check_username_in_url unless @identity_available_called
```
Comment (nessiLahav) [inline app/domain/authentication/authn_jwt/identity_providers/identity_from_url_provider.rb:17]: WDYT to change it to: in this way we will not need an extra boolean 
```
return @identity_available unless @identity_available.nil?
@identity_available ||= check_username_in_url
```

File: app/domain/authentication/authn_jwt/identity_providers/create_identity_provider.rb (hunk 10 above)
Comment (nessiLahav) [reply app/domain/authentication/authn_jwt/identity_providers/create_identity_provider.rb:28]: we want to throw the same error as they both configured

File: app/domain/authentication/authn_jwt/identity_providers/create_identity_provider.rb [hunk 13]
```diff
@@ -22,26 +22,27 @@ def call
         private
 
         def create_identity_provider
-          @logger.debug(LogMessages::Authentication::AuthnJwt::SelectingIdentityProviderInterface.new)
```
Comment (nessiLahav) [inline app/domain/authentication/authn_jwt/identity_providers/create_identity_provider.rb:25]: why did you delete this message?

**Review by Avi-Eyal — APPROVED** [review]

File: app/domain/authentication/authn_jwt/identity_providers/create_identity_provider.rb [hunk 14]
```diff
@@ -23,26 +23,32 @@ def call
 
         def create_identity_provider
           @logger.debug(LogMessages::Authentication::AuthnJwt::SelectingIdentityProviderInterface.new)
-          identity_provider = @identity_from_decoded_token_class.new(@authentication_parameters)
-          if identity_provider.identity_available?
+          identity_from_decoded_token_provider = @identity_from_decoded_token_class.new(@authentication_parameters)
```
Comment (nessiLahav) [inline app/domain/authentication/authn_jwt/identity_providers/create_identity_provider.rb:26]: Please convert `identity_from_decoded_token_provider` & `identity_from_url_provider` to functions with memoizations

**Review by nessiLahav — CHANGES_REQUESTED** [review]
Left 1 comment,
ALso did you add tests to test plan ?

File: app/domain/authentication/authn_jwt/identity_providers/create_identity_provider.rb [hunk 15]
```diff
@@ -21,28 +21,42 @@ def call
...
+
+        def identity_from_url_provider
+          @identity_from_url_provider ||= @identity_from_url_provider_class.new(@authentication_parameters)
+        end
+
         def create_identity_provider
           @logger.debug(LogMessages::Authentication::AuthnJwt::SelectingIdentityProviderInterface.new)
-          identity_provider = @identity_from_decoded_token_class.new(@authentication_parameters)
-          if identity_provider.identity_available?
+          @identity_from_decoded_token_provider = identity_from_decoded_token_provider
```
Comment (nessiLahav) [inline app/domain/authentication/authn_jwt/identity_providers/create_identity_provider.rb:34]: why do we need those lines ?

File: app/domain/authentication/authn_jwt/identity_providers/create_identity_provider.rb [hunk 16]
```diff
@@ -21,28 +21,42 @@ def call
...
+          @identity_from_url_provider = identity_from_url_provider
+
+          check_identity_configuration
+          if @identity_from_decoded_token_provider.identity_available
```
Comment (nessiLahav) [inline app/domain/authentication/authn_jwt/identity_providers/create_identity_provider.rb:38]: use the function instead of the parameter

File: app/domain/authentication/authn_jwt/identity_providers/create_identity_provider.rb [hunk 17]
```diff
@@ -21,28 +21,42 @@ def call
...
                 URL_IDENTITY_PROVIDER_INTERFACE_NAME
               )
             )
-            return identity_provider
+            return @identity_from_url_provider
+          end
+        end
+
+        def check_identity_configuration
+          if (@identity_from_decoded_token_provider.identity_available and @identity_from_url_provider.identity_available) ||
```
Comment (nessiLahav) [inline app/domain/authentication/authn_jwt/identity_providers/create_identity_provider.rb:56]: also here
